def HELP(){
  println '''
Parameter | Default | Description

--pubdir | /<PATH> | The directory that the saved outputs will be stored.
--organize_by | sample | How to organize the output folder structure. Options: sample or analysis
--cacheDir | /panfs/pfs.local/work/sjmac/observer/containers | This is directory that contains cached Singularity containers. Containers are in publicly available repositorys and will be downloaded if not present in this directory at runtime.
-w | /<PATH> | The directory that Nextflow processes use to stage files and intermediate output. This directory can become quite large. This should be a location on scratch space or other directory with ample storage.

--sample_folder | /<PATH> | The path to the folder that contains all the samples to be run by the pipeline. The files in this path can also be symbolic links. 
--extension | .fastq | The expected extension for the input read files.
--pattern | '*_R{1,2}*' | The expected R1 / R2 matching pattern. The default value will match reads with names like this READ_NAME_R1_MoreText.fastq.gz or READ_NAME_R1.fastq.gz
--read_type | SE | Options: PE and SE. Default: SE. Type of reads: paired end (PE) or single end (SE).

--quality_phred | 15 | fastp quality score threshold
--unqualified_perc | 40 | fastp threshold for percent of unqualified bases to pass reads

--fasta | null | Reference FASTA file for alignment and variant identification.
--fasta_fai | null | Index of reference FASTA. Optional argument, if not provided it will be generated by the pipeline.
--bwa_index | null | BWA indices used for read alignment. Optional argument, if not provided it will be generated by the pipeline.

--bwa_min_score | null | bwa mem minimum alignment threshold score
--mismatch_penalty | -B 8 | bwa mem mismatch penalty
--ploidy | 2 | bcftools ploidy value for variant calling (1 or 2)
--mpileup_depth | 100 | For a given position, read maximally INT reads per input file
--skip_indels | false | If true, indels will not be called
--variants_only | true | If true, the output file will not include non-variant sites (i.e. all homozygous reference calls)
--filter_dp | DP < 25 | GATK variant filtering INFO depth threshold
--filter_very_low_qual | QUAL < 30.0 | GATK variant filtering INFO quality threshold (very low)
--filter_low_qual | QUAL > 30.0 && QUAL < 50.0 | GATK variant filtering INFO quality threshold (low)
--filter_qd | QD < 1.5 | GATK variant filtering INFO depth-normalized quality threshold
--filter_fs | FS > 60.0 | GATK variant filtering INFO strand bias threshold

--keep_intermediate | false | Controls whether intermediate alignment files are output.
--keep_reference | false | Controls whether reference indicies generated by pipeline are saved to output directory.
'''
}
